### ISSUE TITLE
Discourse API Keys

### ISSUE THREAD

**Issue by sajattack** [hello-rust/hello#6]
Hey, just wanted to let you know discourse actually does support API keys for non-admins, it's just really poorly documented. There's some documentation [here](https://meta.discourse.org/t/generating-user-api-keys-with-rest-api/61916/13) and [here](https://meta.discourse.org/t/user-api-keys-specification/48536/2) and some code [here](https://github.com/sajattack/discourse-tui/blob/84a4f1c892a002cb97399dcd197d043e04c684a7/src/api.rs#L352).

**Comment by mre** [comment]
Oh wow, that's great news. 😄
The second link mentions that write tokens need to be unlocked by the admins. Can you confirm that?
Would mostly be interested in announcing new episodes of "Hello Rust" on the Rust forum. Have you tried that, perhaps? (Was just thinking because your code is also written in Rust)

**Comment by sajattack** [comment]
Yeah I think I successfully posted to the "What are you working on this week" thread with it if I recall correctly.

**Comment by sajattack** [comment]
Yeah I just made another test post with it and it worked.

**Comment by sajattack** [comment]
I think you have to strip the %0A's

**Comment by sajattack** [comment]
Actually, I just clicked the URL and it loaded up for me. Maybe it was a temporary outage?

**Comment by mre** [comment]
It's strange. The above link works for me now as well and it takes me to the login screen.
Recreated a new link and it doesn't work anymore (Error 500):

https://users.rust-lang.org/user-api-key/new?scopes=read,write&client_id=1ce9289742e7983713e558ac2c263ee4&nonce=714d3c3281d0595a&auth_redirect=discourse://auth_redirect&application_name=Discourse+TUI&public_key=-----BEGIN+PUBLIC+KEY-----MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAznh7RrdsmFtTc/0TvdmC7gyl32aP36H932I0Lt1hOo5sMNyVv8Tr1Q0sgCircIoNhBy+D1OabWoWQHc7xWZ29ih+V2gfl3WNyfRasktZPFjMKpqSeq7Q7Tfben7WEq1vmq1EdQ80bv3vN/fHdGcbeI2aPtBhuyjyAw8ok3chVXcrp5zsnseGpaQeBIrrh2A49oU1dUmMB8BTbtQ/eJkKJJoMsI+XkbfJmgU7mmd3048Tu/y+MupzSdo5IJ4RvRpI/q+vTwD8CDJ/yHjwXZ0remxdF0XFw/OD4Cnz1Qx+D/K8hKlaFldHLZY5tIn7a9cfyrdQahd8D9BDHLWixDj8xQIDAQAB-----END+PUBLIC+KEY-----

Maybe this link will also begin to work after a while? Maybe it's a backend job that needs to get triggered or something? No clue. 🤔

**Comment by mre** [comment]
16 hours in, second link still dead... 😕

**Comment by mre** [comment]
@sajattack, does the second link that I posted work for you? On my machine it doesn't while the first one works now.

**Comment by sajattack** [comment]
No it doesn't. I didn't mean that you should remove all the URL encoding.

**Comment by mre** [comment]
Oh. You're right. Funnily I was just rerunning the tool and then pasting the result here.
I could swear that I copied the link over verbatim but yeah, it works now. 😄

**Comment by mre** [comment]
Thanks so far. 👍 
I guess the next step would be to register the app as a URI handler for `discourse://auth_redirect` as you did [here](https://github.com/sajattack/discourse-tui/tree/84a4f1c892a002cb97399dcd197d043e04c684a7#setup-api-key).
Would like to do that cross-platform (Mac/Linux). Would be cool to have a separate crate for that, but I guess that's out of the scope for now.
Will do that when I find some time (or you'd like to do it, which would also be awesome, of course).

**Comment by mre** [comment]
Whelp, turns out there's https://github.com/maidsafe/system_uri, which we could use. 🎉

**Comment by sajattack** [comment]
Nifty.

**Comment by mre** [comment]
One question: does the authentication redirect really need to point to a `discourse://` URL?
Was searching for this information but couldn't find it.

I'm asking because the system uri handler is quite tricky to get right - especially across platforms. https://github.com/maidsafe/system_uri helps, but one would still have to wrap *hello* into an app on Mac for example.
If we could use a simple `http://` URL as an authentication redirect, things might become a little easier. In this case we could just spawn a little local webserver, open the generated auth link in a normal webbrowser and wait until the user signs in. Then the browser would get redirected to our little webserver which prints the auth token.
Semi-automatic, but easier to set up and doesn't modify the system's uri handlers.

What do you think?

@m3t0r and me were experimenting with this, but we didn't get it to work. See #7 for the current status.

**Comment by sajattack** [comment]
The discourse:// URI is allowed by default. Other redirects have to be enabled by server admins.

**Comment by mre** [comment]
Ah okay. That's unfortunate. Then I guess we'll have to go the hard way...
Thanks for the info! 🤓
